src/mattack_actors.cpp

```diff
@@ -442,6 +451,214 @@ Creature *melee_actor::find_target( monster &z ) const
+        }
+        // The monster might drag a target that's not on it's z level
+        // So if they leave them on open air, make them fall
+        here.creature_on_trap( *target );
+
+        target->add_msg_player_or_npc( msg_type, grab_data.pull_msg_u, grab_data.pull_msg_npc, mon_name,
+                                       target->disp_name() );
+    } else if( grab_data.pull_chance > -1 ) {
+        // We failed the pull chance roll, return false to select a different attack
+        add_msg_debug( debugmode::DF_MATTACK, "Pull chance roll failed.", grab_data.pull_chance );
+        return -1;
+    }
+
+    if( grab_data.grab_effect != effect_null ) {
+        if( foe ) {
+            z.add_effect( bp_id->grabbing_effect, 1_days, true, 1 );
+            add_msg_debug( debugmode::DF_MATTACK, "Added grabbing filter effect %s",
+                           bp_id->grabbing_effect.c_str() );
+            // Add grabbed - permanent, removal handled in try_remove_grab on move/wait
+            target->add_effect( grab_data.grab_effect, 1_days, bp_id, true, eff_grab_strength );
+        } else {
+            // Monsters don't have limb scores, no need to target limbs
+            target->add_effect( grab_data.grab_effect, 1_days, body_part_bp_null, true, eff_grab_strength );
+            z.add_effect( effect_grabbing, 1_days, true, 1 );
+        }
+    }
+
+    // Drag stuff
+    if( grab_data.drag_distance > 0 ) {
+        int distance = grab_data.drag_distance;
+        while( distance > 0 ) {
+            // Start with the opposite square
+            tripoint opposite_square = z.pos() - ( target->pos() - z.pos() );
+            // Keep track of our neighbors (no leaping)
+            std::set<tripoint> neighbors;
+            for( const tripoint &trp : here.points_in_radius( z.pos(), 1 ) ) {
+                if( trp != z.pos() && trp != target->pos() ) {
+                    neighbors.insert( trp );
+                }
+            }
+            // Check where we get to consider dragging
+            std::set<tripoint> candidates;
+            for( const tripoint &trp : here.points_in_radius( opposite_square,
+                    grab_data.drag_deviation ) ) {
+                if( trp != z.pos() && trp != target->pos() ) {
+                    candidates.insert( trp );
+                }
+            }
+            // Select a random square from the options
+            std::set<tripoint> intersect;
+            std::set_intersection( neighbors.begin(), neighbors.end(), candidates.begin(), candidates.end(),
+                                   std::inserter( intersect, intersect.begin() ) );
+            std::set<tripoint>::iterator intersect_iter = intersect.begin();
+            std::advance( intersect_iter, rng( 0, intersect.size() - 1 ) );
+            tripoint target_square = random_entry<std::set<tripoint>>( intersect );
+            if( z.can_move_to( target_square ) ) {
+                monster *zz = dynamic_cast<monster *>( target );
```

Comment:
```suggestion
                monster *zz = target->as_monster();
```

Also `target->is_monster()` and so on is nicer and clearer than nullptr checks.